File: java/Components/Strata.Domain.InMemory/src/strata/domain/inmemory/InMemoryUnitOfWork.java
// ##########################################################################
// # File Name:	InMemoryUnitOfWork.java
// #
// # Copyright:	2012, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataEntity Framework.
// #
// #   			The StrataEntity Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataEntity Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataEntity
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.domain.inmemory;

import strata.domain.core.namedquery.INamedQuery;
import strata.domain.core.repository.IKeyRetriever;
import strata.domain.core.repository.InsertFailedException;
import strata.domain.core.repository.RemoveFailedException;
import strata.domain.core.repository.UpdateFailedException;
import strata.domain.core.unitofwork.AbstractUnitOfWork;
import strata.domain.core.unitofwork.CommitFailedException;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutorService;


/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class InMemoryUnitOfWork
    extends AbstractUnitOfWork
{
    private Map<EntityIdentifier,Object> itsInserted;
    private Map<EntityIdentifier,Object> itsUpdated;
    private Map<EntityIdentifier,Object> itsRemoved;
    private Map<EntityIdentifier,Object> itsEntities;
    private final ExecutorService        itsExecutor;
    
    /************************************************************************
     * Creates a new {@code InMemoryUnitOfWork}. 
     *
     */
    public 
    InMemoryUnitOfWork(
        InMemoryUnitOfWorkProvider   provider,
        Map<EntityIdentifier,Object> entities)
    {
        super( provider );
        itsInserted = new HashMap<>();
        itsUpdated  = new HashMap<>();
        itsRemoved  = new HashMap<>();
        itsEntities = entities;
        itsExecutor = provider.getExecutor();
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    close()
    {
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> CompletionStage<E>
    doInsertNew(Class<K> keyType,Class<E> entityType,E newEntity)
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    System.out.println("InMemoryUnitOfWork.doInsertNew: " + Thread.currentThread().getName());
                    E                insertedEntity = null;
                    K                key            = null;
                    EntityIdentifier id             = null;

                    try
                    {
                        insertedEntity =
                            getReplicator(keyType,entityType).replicate(newEntity);
                        key =
                            getRetriever(keyType,entityType).getKey(insertedEntity);
                        id = new EntityIdentifier(entityType,key);

                        if (itsUpdated.containsKey(id))
                            throw new IllegalArgumentException("entity is already updated");

                        if (itsRemoved.containsKey(id))
                            throw new IllegalArgumentException("entity is already removed");

                        if (
                            itsInserted.containsKey(id) ||
                                doHasExistingInternal(entityType,key))
                            throw new IllegalArgumentException("entity is already inserted");

                        itsInserted.put(id,insertedEntity);
                        return insertedEntity;
                    }
                    catch (Exception e)
                    {
                        throw
                            new CompletionException(
                                new InsertFailedException(e));
                    }

                },
                itsExecutor);

    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> CompletionStage<E>
    doUpdateExisting(Class<K> keyType,Class<E> entityType,E existingEntity)
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    System.out.println("InMemoryUnitOfWork.doUpdateExisting: " + Thread.currentThread().getName());
                    E                updatedEntity = null;
                    K                key           = null;
                    EntityIdentifier id            = null;

                    try
                    {
                        updatedEntity =
                            getReplicator(keyType,entityType).replicate(existingEntity);
                        key =
                            getRetriever(keyType,entityType).getKey(updatedEntity);
                        id = new EntityIdentifier(entityType,key);

                        if (itsRemoved.containsKey(id))
                            throw new IllegalArgumentException("entity is already removed");

                        if (
                            !itsInserted.containsKey(id) &&
                                !doHasExistingInternal(entityType,key))
                            throw new IllegalArgumentException("entity does not exist");

                        itsUpdated.put(new EntityIdentifier(entityType,key),updatedEntity);
                        return updatedEntity;
                    }
                    catch (Exception e)
                    {
                        throw
                            new CompletionException(
                                new UpdateFailedException(e));
                    }
                },
                itsExecutor);
    }

    /************************************************************************
     * {@inheritDoc}
     * @return
     */
    @Override
    protected <K extends Serializable,E> CompletionStage<E>
    doRemoveExisting(Class<K> keyType,Class<E> entityType,E existingEntity)
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    K                key = null;
                    EntityIdentifier id  = null;

                    try
                    {
                        key =
                            getRetriever(keyType,entityType).getKey(existingEntity);
                        id = new EntityIdentifier(entityType,key);

                        if (
                            !itsInserted.containsKey(id) &&
                                !doHasExistingInternal(entityType,key))
                            throw new IllegalArgumentException("entity does not exist");

                        itsRemoved.put(id,existingEntity);
                        return existingEntity;
                    }
                    catch (Exception e)
                    {
                        throw
                            new CompletionException(
                                new RemoveFailedException(e));
                    }

                },
                itsExecutor);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> CompletionStage<Optional<E>>
    doGetExisting(Class<E> type,K key)
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    System.out.println("InMemoryUnitOfWork.doGetExisting: " + Thread.currentThread().getName());
                    EntityIdentifier id = new EntityIdentifier(type,key);

                    if ( itsRemoved.containsKey( id ) )
                        return Optional.empty();

                    if ( itsUpdated.containsKey( id ) )
                        return
                            Optional.ofNullable(
                                type.cast(itsUpdated.get(id)));

                    if ( itsInserted.containsKey( id ) )
                        return
                            Optional.ofNullable(
                                type.cast(itsInserted.get(id)));

                    if ( itsEntities.containsKey( id ) )
                        return
                            Optional.ofNullable(
                                type.cast(itsEntities.get(id)));

                    return
                        Optional.ofNullable(
                            getCandidates(key.getClass(),type).get(key));

                },
                itsExecutor);

    }

    /************************************************************************
     * {@inheritDoc}
     * @return
     */
    @Override
    protected <E> CompletionStage<INamedQuery<E>>
    doGetNamedQuery(Class<E> type,String queryName)
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    INamedQuery<E> query =
                        ((InMemoryUnitOfWorkProvider)getProvider())
                            .getNamedQuery(type,queryName);

                    if ( query == null )
                        throw new NullPointerException(queryName + " not found");

                    return query;
                },
                itsExecutor);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> CompletionStage<Boolean>
    doHasExisting(Class<E> type,K key)
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    System.out.println("InMemoryUnitOfWork.doHasExisting: " + Thread.currentThread().getName());
                    EntityIdentifier id = new EntityIdentifier(type,key);

                    if ( itsRemoved.containsKey( id ) )
                        return false;

                    if ( itsUpdated.containsKey( id ) )
                        return true;

                    if ( itsInserted.containsKey( id ) )
                        return true;

                    if ( itsEntities.containsKey( id ) )
                        return true;

                    return
                        getCandidates(key.getClass(),type)
                            .containsKey(key);

                },
                itsExecutor);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <E> CompletionStage<Boolean>
    doHasNamedQuery(Class<E> type,String queryName)
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    System.out.println("InMemoryUnitOfWork.doHasNamedQuery: " + Thread.currentThread().getName());
                    return doGetNamedQuery(type,queryName) != null;
                },
                itsExecutor);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected CompletionStage<Void>
    doCommit()
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    synchronized (itsEntities)
                    {
                        try
                        {
                            commitInserted();
                            commitUpdated();
                            commitRemoved();
                            return null;
                        }
                        catch (Exception cause)
                        {
                            throw
                                new CompletionException(
                                    new CommitFailedException(cause));
                        }
                    }
                },
                itsExecutor);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected CompletionStage<Void>
    doRollback()
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    itsInserted.clear();
                    itsUpdated.clear();
                    itsRemoved.clear();
                    return null;
                },
                itsExecutor);
    }


    /************************************************************************
     *  
     *
     */
    protected void
    commitInserted()
    {
        for (
            Map.Entry<EntityIdentifier,Object> entry :
                itsInserted.entrySet())
            itsEntities.put( entry.getKey(),entry.getValue() );
    }
    
    /************************************************************************
     *  
     *
     */
    protected void
    commitUpdated()
    {
        for (
            Map.Entry<EntityIdentifier,Object> entry :
                itsUpdated.entrySet())
            itsEntities.put( entry.getKey(),entry.getValue() );
    }
   
    /************************************************************************
     *  
     *
     */
    protected void
    commitRemoved()
    {
        for (
            Map.Entry<EntityIdentifier,Object> entry :
                itsRemoved.entrySet())
        {
            if ( itsEntities.remove( entry.getKey() ) == null )
                throw new IllegalStateException("entity does not exist");
        }
    }

    /************************************************************************
     *  
     *
     * @param keyType
     * @param entityType
     * @return
     */
    private <K extends Serializable,E> IEntityReplicator<K,E>
    getReplicator(Class<K> keyType,Class<E> entityType)
    {
        return 
            ((InMemoryUnitOfWorkProvider)getProvider())
                .getReplicator( keyType,entityType );
    }

    /************************************************************************
     *  
     *
     * @param keyType
     * @param entityType
     * @return
     */
    private <K extends Serializable,E> IKeyRetriever<K,E>
    getRetriever(Class<K> keyType,Class<E> entityType)
    {
        return 
            ((InMemoryUnitOfWorkProvider)getProvider())
                .getRetriever( keyType,entityType );
    }

    /************************************************************************
     *  
     *
     * @param keyType
     * @param entityType
     * @return
     */
    private <K,T> Map<K,T>
    getCandidates(Class<K> keyType,Class<T> entityType)
    {
        Map<K,T> candidates = new HashMap<K,T>();
        
        for (
            Map.Entry<EntityIdentifier,Object> entry:
                itsEntities.entrySet())
        {
            if ( entityType.isAssignableFrom( entry.getKey().getType() ) )
            {
                try
                {
                    candidates.put( 
                        keyType.cast( entry.getKey().getKey() ),
                        entityType.cast( entry.getValue() ) );
                }
                catch (ClassCastException e) {}
            }
        }
        
        return candidates;
    }


    /************************************************************************
     * {@inheritDoc}
     */
    protected <E,K> boolean
    doHasExistingInternal(Class<E> type,K key)
    {
        EntityIdentifier id = new EntityIdentifier(type,key);

        if ( itsRemoved.containsKey( id ) )
            return false;

        if ( itsUpdated.containsKey( id ) )
            return true;

        if ( itsInserted.containsKey( id ) )
            return true;

        if ( itsEntities.containsKey( id ) )
            return true;

        return
            getCandidates(key.getClass(),type)
                .containsKey(key);
    }

}

// ##########################################################################

File: csharp/Components/Strata.Domain.InMemory/UnitOfWOrk/InMemoryUnitOfWork.cs
//  ##########################################################################
//  # File Name: InMemoryUnitOfWork.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Strata.Domain.Core.NamedQuery;
using Strata.Domain.Core.UnitOfWork;
using Strata.Foundation.Core.Utility;

namespace Strata.Domain.InMemory.UnitOfWOrk
{

    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public
    class InMemoryUnitOfWork:
        AbstractUnitOfWork
    {
        public bool IsDisposed { get; protected set; }

        private readonly IDictionary<EntityIdentifier,Object>  inserted;
        private readonly IDictionary<EntityIdentifier,Object>  updated;
        private readonly IDictionary<EntityIdentifier,Object>  removed;
        private readonly IDictionary<EntityIdentifier,Object>  entities;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>InMemoryUnitOfWork</c> instance.
        /// </summary>
        /// 
        /// <param name="ent">in memory entities</param>
        /// 
        public 
        InMemoryUnitOfWork(
            InMemoryUnitOfWorkProvider           provider,
            IDictionary<EntityIdentifier,Object> ent):
            base(provider)
        {
            inserted   = new Dictionary<EntityIdentifier,Object>();
            updated    = new Dictionary<EntityIdentifier,Object>();
            removed    = new Dictionary<EntityIdentifier,Object>();
            entities   = ent;
            IsDisposed = false;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void
        Dispose()
        {
            inserted.Clear();
            updated.Clear();
            removed.Clear();
            IsDisposed = true;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override async Task<E> 
        DoInsert<K,E>(E entity)
        {
            E                insertedEntity = null;
            K                key            = default(K);
            EntityIdentifier id             = null;
        
            insertedEntity = GetReplicator<E>().Replicate(entity);
            key            = GetRetriever<K,E>()(insertedEntity);
            id             = new EntityIdentifier(typeof(E),key);
        
            if ( updated.ContainsKey( id ) )
                throw new ArgumentException("entity is already updated");
        
            if ( removed.ContainsKey( id ) )
                throw new ArgumentException("entity is already removed");
        
            if ( inserted.ContainsKey(id) || await DoHasUnique<K,E>(key))
                throw new ArgumentException("entity is already inserted");
        
            inserted.Add( id,insertedEntity );
            return insertedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override async Task<E> 
        DoUpdate<K,E>(E entity)
        {
            E                updatedEntity = null;
            K                key           = default(K);
            EntityIdentifier id            = null;
            bool             hasEntity     = false;
        
            updatedEntity = GetReplicator<E>().Replicate(entity);
            key =  GetRetriever<K,E>()(updatedEntity);
            id = new EntityIdentifier(typeof(E),key);
        
            if ( removed.ContainsKey( id ) )
                throw new ArgumentException("entity is already removed");

            hasEntity = await DoHasUnique<K,E>(key);

            if ( !inserted.ContainsKey(id) && !hasEntity)
                throw new ArgumentException("entity does not exist");
        
            updated.Add( id,updatedEntity );
            return updatedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override async Task
        DoRemove<K,E>(E entity)
        {
            K                key = default(K);
            EntityIdentifier id  = null;
        
            key = GetRetriever<K,E>()(entity);
            id = new EntityIdentifier(typeof(E),key);

            if ( !inserted.ContainsKey(id) && !await DoHasUnique<K,E>(key))
                throw new ArgumentException("entity does not exist");
        
            removed.Add( id,entity ); 
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<Optional<E>> 
        DoGetUnique<K,E>(K key)
        {
            EntityIdentifier id = new EntityIdentifier(typeof(E),key);
        
            if ( removed.ContainsKey( id ) )
                return Task.FromResult(Optional<E>.Empty());
 
            if ( updated.ContainsKey( id ) )
                return 
                    Task.FromResult(
                        Optional<E>.OfNullable((E)updated[id]));

            if ( inserted.ContainsKey( id ) )
                return 
                    Task.FromResult(
                        Optional<E>.OfNullable((E)inserted[id]));
        
            return 
                Task.FromResult(
                    Optional<E>.OfNullable(
                        GetReplicator<E>()
                            .Replicate(GetCandidates<K,E>()[key])));
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<Optional<E>> 
        DoGetUniqueMatching<E>(Expression<Func<E,bool>> predicate)
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);
            E result =
                queryable
                    .Values
                    .OfType<E>()
                    .AsQueryable()
                    .Where(predicate)
                    .SingleOrDefault();

            if (result != null)
                return 
                    Task.FromResult(
                        Optional<E>.OfNullable(
                            GetReplicator<E>().Replicate(result)));

            queryable = new Dictionary<EntityIdentifier,object>();
            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            return
                Task.FromResult(
                    Optional<E>.OfNullable(
                        queryable
                            .Values
                            .OfType<E>()
                            .AsQueryable()
                            .Where(predicate)
                            .SingleOrDefault()));
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<IList<E>> 
        DoGetMatching<E>(Expression<Func<E,bool>> predicate)
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);
            IList<E> temp =
                queryable
                    .Values
                    .OfType<E>()
                    .AsQueryable()
                    .Where(predicate)
                    .ToList();
            IList<E> results = new List<E>();

            if (temp != null && temp.Count > 0)
                foreach (E entity in temp)
                    results.Add(GetReplicator<E>().Replicate(entity));

            queryable = new Dictionary<EntityIdentifier,object>();
            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            results.AddAll(
                queryable
                    .Values
                    .OfType<E>()
                    .AsQueryable()
                    .Where(predicate)
                    .ToList());

            return Task.FromResult(results);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<IList<E>> 
        DoGetAll<E>()
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);
            IList<E> temp =
                queryable
                    .Values
                    .OfType<E>()
                    .ToList();
            IList<E> results = new List<E>();

            if (temp != null && temp.Count > 0)
                foreach (E entity in temp)
                    results.Add(GetReplicator<E>().Replicate(entity));

            queryable = new Dictionary<EntityIdentifier,object>();
            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            results.AddAll(
                queryable
                    .Values
                    .OfType<E>()
                    .ToList());

            return Task.FromResult(results);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<INamedQuery<E>> 
        DoGetNamedQuery<E>(string queryName)
        {
            return
                Task.FromResult(
                    ((InMemoryUnitOfWorkProvider) Provider)
                        .GetNamedQuery<E>(queryName));
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<bool> 
        DoHasUnique<K,E>(K key)
        {
            EntityIdentifier id = new EntityIdentifier(typeof(E),key);
        
            if ( removed.ContainsKey( id ) )
                return Task.FromResult(false);
 
            if ( updated.ContainsKey( id ) )
                return Task.FromResult(true);

            if ( inserted.ContainsKey( id ) )
                return Task.FromResult(true);
        
            return Task.FromResult(GetCandidates<K,E>().ContainsKey(key));
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<bool> 
        DoHasMatching<E>(Expression<Func<E,bool>> predicate)
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);

            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            return
                Task.FromResult(
                    queryable
                        .Values
                        .OfType<E>()
                        .AsQueryable()
                        .Where(predicate)
                        .Any());
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<bool> 
        DoHasNamedQuery<E>(string queryName)
        {
            return 
                Task.FromResult(
                    ((InMemoryUnitOfWorkProvider)Provider)
                        .HasNamedQuery<E>(queryName));
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task 
        DoCommit()
        {
            Task task =
                new Task(
                    () =>
                    {
                        try
                        {
                            CommitInserted();
                            CommitUpdated();
                            CommitRemoved();

                            inserted.Clear();
                            updated.Clear();
                            removed.Clear();
                        }
                        catch (Exception cause)
                        {
                            throw new CommitFailedException("commit failed",cause);
                        }
                    });
            
            task.Start();
            return task;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override async Task 
        DoRollback()
        {
            try
            {
                inserted.Clear();
                updated.Clear();
                removed.Clear();
                await base.DoRollback();
            }
            catch (RollbackFailedException)
            {
                throw;
            }
            catch (Exception cause)
            {
                throw new RollbackFailedException("rollback failed",cause);
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public E 
        DoInsert<K,E>(
            E                    entity,
            IEntityReplicator<E> replicator,
            KeyRetriever<K,E>    retriever)
            where E:class
        {
            E                insertedEntity = replicator.Replicate( entity );
            K                key = retriever( insertedEntity );
            EntityIdentifier id  = new EntityIdentifier(typeof(E),key);   
     
            if ( updated.ContainsKey( id ) )
                throw new InvalidOperationException("entity is already updated");
        
            if ( removed.ContainsKey( id ) )
                throw new InvalidOperationException("entity is already removed");
        
            if ( inserted.ContainsKey( id ) || DoHas<K,E>( key ) )
                throw new InvalidOperationException("entity is already inserted");
        
            inserted.Add( id,insertedEntity );
            return insertedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public  E 
        DoUpdate<K,E>(
            E                    entity,
            IEntityReplicator<E> replicator,
            KeyRetriever<K,E>    retriever)
            where E:class
        {
            E                updatedEntity = replicator.Replicate( entity );
            K                key = retriever( updatedEntity );
            EntityIdentifier id  = new EntityIdentifier(typeof(E),key);   
        
            if ( !inserted.ContainsKey( id ) && !DoHas<K,E>( key ) )
                throw new InvalidOperationException("entity does not exist");
        
            if ( removed.ContainsKey( id ) )
                throw new InvalidOperationException("entity is already removed");
        
            if ( updated.ContainsKey( id ) )
                updated[id] = updatedEntity;
            else
                updated.Add( id,updatedEntity );

            return updatedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void 
        DoRemove<K,T>(T entity,KeyRetriever<K,T> retriever)
        {
            K                key = retriever( entity );
            EntityIdentifier id  = new EntityIdentifier( typeof(T),key );       
         
            if ( !inserted.ContainsKey( id ) && !DoHas<K,T>( key ) )
                throw new InvalidOperationException("entity does not exist");

            removed.Add( id,entity );
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public T
        DoGet<K,T>(K key)
            where T:class
        {
            EntityIdentifier id = new EntityIdentifier( typeof(T),key );

            if ( removed.ContainsKey( id ) )
                return null;

            if ( updated.ContainsKey( id ) )
                return (T)updated[id];

            if ( inserted.ContainsKey( id ) )
                return (T)inserted[id];

            return (T)entities[id];
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public IQueryable<T>
        DoQuery<T>()
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier, Object>(entities);

            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

             return
                queryable
                    .Values
                    .OfType<T>()
                    .AsQueryable();
        }
    
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public bool
        DoHas<K,T>(K key)
        {
            EntityIdentifier id = new EntityIdentifier( typeof(T),key );

            if ( removed.ContainsKey( id ) )
                return false;

            if ( inserted.ContainsKey( id ) )
                return true;

            if ( updated.ContainsKey( id ) )
                return true;

            return entities.ContainsKey( id );
        }
    
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void
        CommitInserted()
        {
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry in inserted)
                entities.Add( entry.Key,Replicate(entry.Value) );
        }
    
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void
        CommitUpdated()
        {
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry in updated)
                entities[entry.Key] = Replicate(entry.Value);
        }
   
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void
        CommitRemoved()
        {
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry in removed)
            {
                if ( entities.Remove( entry.Key ) == false )
                    throw 
                        new InvalidOperationException("entity does not exist");
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public IEntityReplicator<T>
        GetReplicator<T>()
            where T: class
        {
            return 
                ((InMemoryUnitOfWorkProvider)Provider).GetReplicator<T>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public KeyRetriever<K,T>
        GetRetriever<K,T>()
            where T: class
        {
            return 
                ((InMemoryUnitOfWorkProvider)Provider).GetRetriever<K,T>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private IDictionary<K,T>
        GetCandidates<K,T>()
            where T: class
        {
            IDictionary<K,T> candidates = new Dictionary<K,T>();
            KeyRetriever<K,T> retriever = GetRetriever<K,T>();

            foreach (T entity in GetEntitiesByType<T>())
                candidates.Add(retriever(entity),entity);
                
            return candidates;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private IList<T>
        GetEntitiesByType<T>()
        {
            Type     type   = typeof(T);
            IList<T> output = new List<T>();
        
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry 
                    in entities)
            {
                if ( type.IsAssignableFrom( entry.Key.EntityType ) )
                    output.Add( (T)( entry.Value ) );
            }
        
            return output;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private object
        Replicate(object source)
        {
            IEntityReplicator<object> replicator = 
                ((InMemoryUnitOfWorkProvider)Provider).GetObjectReplicator(
                    source.GetType());

            return replicator.Replicate(source);
        }
    }
}

//  ##########################################################################
